File: services/auth-service/src/utils/__tests__/password.test.ts
import { describe, it, expect } from 'vitest';
import { hashPassword, comparePassword, validatePasswordStrength } from '../password';

describe('Password Utils', () => {
  describe('validatePasswordStrength', () => {
    it('should accept valid passwords', () => {
      expect(validatePasswordStrength('ValidPass1!')).toBe(true);
      expect(validatePasswordStrength('Test@123456')).toBe(true);
      expect(validatePasswordStrength('MyP@ssw0rd')).toBe(true);
    });

    it('should reject passwords without uppercase', () => {
      expect(validatePasswordStrength('nouppercas3!')).toBe(false);
    });

    it('should reject passwords without lowercase', () => {
      expect(validatePasswordStrength('NOLOWERCASE3!')).toBe(false);
    });

    it('should reject passwords without numbers', () => {
      expect(validatePasswordStrength('NoNumbers!')).toBe(false);
    });

    it('should reject passwords without special characters', () => {
      expect(validatePasswordStrength('NoSpecial1')).toBe(false);
    });

    it('should reject passwords shorter than 8 characters', () => {
      expect(validatePasswordStrength('Short1!')).toBe(false);
    });
  });

  describe('hashPassword and comparePassword', () => {
    it('should hash password and verify correctly', async () => {
      const password = 'TestPassword123!';
      const hash = await hashPassword(password);

      expect(hash).not.toBe(password);
      expect(hash.length).toBeGreaterThan(0);

      const isValid = await comparePassword(password, hash);
      expect(isValid).toBe(true);
    });

    it('should fail verification with wrong password', async () => {
      const password = 'TestPassword123!';
      const wrongPassword = 'WrongPassword123!';
      const hash = await hashPassword(password);

      const isValid = await comparePassword(wrongPassword, hash);
      expect(isValid).toBe(false);
    });
  });
});

File: services/auth-service/pkg/password/password_test.go
package password

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestHashPassword(t *testing.T) {
	plainPassword := "MySecurePass123!"

	hash, err := HashPassword(plainPassword)

	require.NoError(t, err, "Should hash password without error")
	assert.NotEmpty(t, hash, "Hash should not be empty")
	assert.NotEqual(t, plainPassword, hash, "Hash should not equal plain password")
	assert.Greater(t, len(hash), 30, "Hash should be substantial length")
}

func TestComparePassword_CorrectPassword(t *testing.T) {
	plainPassword := "MySecurePass123!"

	hash, err := HashPassword(plainPassword)
	require.NoError(t, err)

	match, err := ComparePassword(plainPassword, hash)

	require.NoError(t, err, "Should compare without error")
	assert.True(t, match, "Correct password should match hash")
}

func TestComparePassword_IncorrectPassword(t *testing.T) {
	correctPassword := "MySecurePass123!"
	wrongPassword := "WrongPassword456!"

	hash, err := HashPassword(correctPassword)
	require.NoError(t, err)

	match, err := ComparePassword(wrongPassword, hash)

	require.NoError(t, err, "Should compare without error")
	assert.False(t, match, "Incorrect password should not match hash")
}

func TestValidatePasswordStrength_ValidPasswords(t *testing.T) {
	validPasswords := []string{
		"ValidPass1!",
		"Str0ng@Password",
		"MyP@ssw0rd123",
		"Test#1234Password",
	}

	for _, password := range validPasswords {
		t.Run(password, func(t *testing.T) {
			assert.True(t, ValidatePasswordStrength(password),
				"Valid password should pass validation: %s", password)
		})
	}
}

func TestValidatePasswordStrength_InvalidPasswords(t *testing.T) {
	testCases := []struct {
		name     string
		password string
		reason   string
	}{
		{
			name:     "no_uppercase",
			password: "nouppercas3!",
			reason:   "missing uppercase letter",
		},
		{
			name:     "no_lowercase",
			password: "NOLOWERCASE3!",
			reason:   "missing lowercase letter",
		},
		{
			name:     "no_digit",
			password: "NoDigitPass!",
			reason:   "missing digit",
		},
		{
			name:     "no_special",
			password: "NoSpecial123",
			reason:   "missing special character",
		},
		{
			name:     "too_short",
			password: "Sh0rt!",
			reason:   "less than 8 characters",
		},
		{
			name:     "empty",
			password: "",
			reason:   "empty password",
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			assert.False(t, ValidatePasswordStrength(tc.password),
				"Invalid password should fail validation: %s (%s)", tc.password, tc.reason)
		})
	}
}

func TestHashPassword_DifferentHashes(t *testing.T) {
	password := "SamePassword123!"

	hash1, err1 := HashPassword(password)
	hash2, err2 := HashPassword(password)

	require.NoError(t, err1)
	require.NoError(t, err2)
	assert.NotEqual(t, hash1, hash2, "Same password should produce different hashes (due to salt)")
}
